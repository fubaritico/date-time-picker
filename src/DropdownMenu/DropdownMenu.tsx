import clsx from 'clsx'
import {
  type FC,
  type ReactNode,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react'

import { Button, Icon, Label, Menu } from '..'

import { isString } from './DropdownMenu.utils'

import type { ButtonProps, MenuItemConfig, MenuPlacement } from '..'

export interface DropdownMenuProps {
  /* Extra CSS classes for the button (tailwind) */
  buttonClassName?: string
  /* The version of button use in place of SimpleButton, with or without inactivity reset time (use for menu items) */
  buttonComponent?: typeof Button
  /* Default label displayed in the opener button */
  buttonLabel?: ReactNode
  /* Button type */
  buttonType?: 'button' | 'submit' | 'reset'
  /* Button color */
  color?: UIColor
  /* Deprecated: Value on init (preselection in the case of controlled component */
  defaultValue?: string
  /* Dropdown is disabled */
  disabled?: boolean
  /* If true will render the menu in a portal */
  enablePortal?: boolean
  /* Array of menu items */
  items?: MenuItemConfig[]
  /* Text input label */
  label?: string
  /* Information displayed on hovering the exclamation mark in a tooltip */
  labelInfo?: string
  /* Loading state (in case values are fetched) */
  loading?: boolean
  /* Menu width in pixels */
  menuWidth?: number
  /* If true, the menu will have the same width as the trigger */
  menuFullWidth?: boolean
  /* If true will render the dropdown in full width */
  dropdownFullWidth?: boolean
  /* Callback called on any menu item click */
  onValueChange?: (value: string) => void
  /* The open state can be controlled */
  open?: boolean
  /* Menu placement: 'bottom-start' | 'bottom-end'  */
  position?: MenuPlacement
  /* Dropdown is required */
  required?: boolean
  /* Extra CSS classes for the root element (tailwind) */
  rootClassName?: string
  /* Button shape 'rounded' | 'square' */
  shape?: 'rounded' | 'square'
  /*
   * Will show the selected value instead of the button label, defaults to false.
   * Use this if you want a selectable menu item to be highlighted when selected.
   */
  showSelectedValue?: boolean
  /* Dropdown button size 'small' | 'medium' | 'large' */
  size?: UISize
  /* Placement of the tooltip displayed on menu item hovering */
  Placement?: Placement
  /* Component used instead of the button, defaults to null */
  triggerComponent?: ReactNode
  /* Extra CSS classes for the trigger component element (tailwind) */
  triggerComponentClassName?: string
  /* Value passed to the component when controlled */
  value?: string
  /* Menu button style 'primary' | 'secondary' */
  variant?: ButtonProps['variant']
  /* Width of the dropdown button in pixels */
  width?: number
}

const labelSizes: Record<UISize, UISize> = {
  small: 'small',
  medium: 'medium',
  large: 'medium',
}

const DropdownMenu: FC<DropdownMenuProps> = ({
  variant = 'primary',
  buttonComponent,
  buttonLabel,
  buttonType,
  disabled,
  enablePortal = false,
  items,
  label,
  labelInfo,
  menuWidth,
  menuFullWidth,
  dropdownFullWidth = false,
  onValueChange,
  open,
  position = 'bottom-end',
  required,
  rootClassName,
  shape = 'rounded',
  showSelectedValue,
  size = 'large',
  triggerComponent = null,
  triggerComponentClassName,
  value,
  width,
  loading,
}) => {
  const [isOpen, setIsOpen] = useState(open ?? false)
  const buttonContainerRef = useRef<HTMLDivElement>(null)
  const currentLabel =
    showSelectedValue && value
      ? items?.find((i) => i.value === value)?.label
      : (buttonLabel ?? '')

  /**
   *
   */
  const closeMenu = useCallback(() => {
    setIsOpen(false)
  }, [])

  /**
   * Will emit the value of a selectable menu item
   *
   * @param item
   */
  const handleMenuItemClick = (item: MenuItemConfig) => {
    if (item.type === 'selectable' && !item.value) {
      throw Error('Value is required for selectable menu item')
    }

    if (isString(item.label) && item.type === 'selectable' && !!item.value) {
      onValueChange?.(item.value)
    }

    if (item.type !== 'selectable') {
      // Will unselect any selectable menu item already selected
      onValueChange?.('')
    }

    closeMenu()
  }

  useEffect(() => {
    document
      .getElementById('page-content')
      ?.addEventListener('scroll', closeMenu)

    return () => {
      document
        .getElementById('page-content')
        ?.removeEventListener('scroll', closeMenu)
    }
  }, [closeMenu])

  return (
    <div
      className={clsx('flex flex-col gap-1', rootClassName, {
        'max-w-[210px]': !dropdownFullWidth,
        'pointer-events-none': disabled ?? loading,
      })}
    >
      {label && (
        <Label
          className="mb-1"
          label={label}
          labelInfo={labelInfo}
          required={required}
          size={labelSizes[size]}
          disabled={disabled}
        />
      )}
      <div className="relative w-full">
        <div
          className={clsx(
            'w-auto',
            {
              '!w-fit': triggerComponent,
            },
            triggerComponentClassName
          )}
          ref={buttonContainerRef}
        >
          {triggerComponent ? (
            <button
              className="appearance-none border-0 bg-transparent p-0"
              onClick={() => {
                setIsOpen(!isOpen)
              }}
              aria-label={isString(buttonLabel) ? buttonLabel : undefined}
            >
              {triggerComponent}
            </button>
          ) : (
            <Button
              aria-label={
                isString(buttonLabel) ? buttonLabel : 'Select an option'
              }
              label={
                <>
                  <span
                    data-test="dropdown-value"
                    className="truncate flex-shrink"
                  >
                    {currentLabel}
                  </span>
                  <div
                    className={clsx('w-5', {
                      'mt-0.5': !isOpen,
                      'mb-0.5': isOpen,
                    })}
                  >
                    <Icon
                      name={isOpen ? 'HiChevronUp' : 'HiChevronDown'}
                      className="size-5"
                    />
                  </div>
                </>
              }
              loading={loading}
              onClick={() => {
                setIsOpen(!isOpen)
              }}
              size={size}
              style={{ width }}
              type={buttonType}
              variant={variant}
            ></Button>
          )}
        </div>
        <Menu
          buttonComponent={buttonComponent}
          className={clsx({ 'w-full': menuFullWidth })}
          enablePortal={enablePortal}
          items={items ?? []}
          menuWidth={menuWidth} // Don't really know what to do with that, to be settled.
          onClose={closeMenu}
          onMenuItemClick={handleMenuItemClick}
          open={isOpen}
          placement={position}
          selectedValue={value}
          shape={shape}
          showSelectedValue={showSelectedValue}
          size={size}
          triggerRef={buttonContainerRef}
        />
      </div>
    </div>
  )
}

export default DropdownMenu
