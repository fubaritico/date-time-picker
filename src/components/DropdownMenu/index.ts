export { default } from './DropdownMenu'

/* Types */
export type { DropdownMenuProps } from './DropdownMenu'
