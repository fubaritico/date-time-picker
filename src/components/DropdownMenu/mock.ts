import type { MenuItemConfig } from '@components'

export const menuItems: MenuItemConfig[] = [
  {
    label: 'Quick save (selectable)',
    value: 'Quick save',
  },
  {
    label: 'Save (action)',
    onButtonClick: () => {
      alert('Save')
    },
    type: 'action',
  },
  {
    label: 'Reassign (action)',
    onButtonClick: () => {
      alert('Reassigning')
    },
    type: 'action',
  },
  {
    label: 'Edit (selectable)',
    value: 'Edit',
  },
  {
    label: 'Delete',
    onButtonClick: () => {
      alert('Delete (action)')
    },
    type: 'action',
  },
]

export const selectableMenuItems: MenuItemConfig[] = [
  {
    label: 'Quick save ',
    value: 'Quick save',
  },
  {
    label: 'Save',
    value: 'Save',
  },
  {
    label: 'Reassign',
    value: 'Reassign',
  },
  {
    label: 'Edit',
    value: 'Edit',
  },
  {
    label: 'Delete',
    value: 'Delete',
  },
]

export const longMenuItems: MenuItemConfig[] = [
  {
    label: 'This is a very long menu item for an "action"',
    onButtonClick: () => {
      alert('This is a very long menu item for an "action"')
    },
  },
  {
    label: 'Another long menu item for an "selectable" item',
    value: 'value2',
  },
  {
    label: 'Very long menu item with icon for an "selectable" item',
    value: 'value3',
    icon: 'HiDocument',
  },
  {
    label: 'Another long menu item for an "selectable" item',
    onButtonClick: () => {
      alert('Another long menu item for an "selectable" item')
    },
  },
  {
    label: 'Very long menu item with icon for an "action" item',
    onButtonClick: () => {
      alert('Very long menu item with icon for an "action" item')
    },
    icon: 'HiDocument',
  },
  {
    label: 'sed do eiusmod tempor incididunt ut labore et dolore magna aliqua',
    value: 'value6',
  },
  {
    label: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
    value: 'value7',
  },
  {
    label: 'adipiscing elit, sed do eiusmod tempor incididunt ut ',
    value: 'value8',
  },
  {
    label: 'ea commodo consequat. Duis aute irure dolor tende',
    value: 'value9',
  },
  {
    label: 'iusmod temea commodo consequat. Duis aute irure d',
    value: 'value10',
  },
]

export const coloredMenuItems: MenuItemConfig[] = [
  {
    label: 'Quick save',
    onButtonClick: () => {
      alert('Quick save')
    },
    icon: 'HiDocument',
  },
  {
    label:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod temea commodo consequat. Duis aute irure dolor tende minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor tende.',
    onButtonClick: () => {
      alert('Save')
    },
  },
  {
    label: 'Reassign',
    onButtonClick: () => {
      alert('Reassign')
    },
    icon: 'HiEnvelope',
    color: 'yellow',
  },
  {
    label: 'Edit',
    onButtonClick: () => {
      alert('Edit')
    },
    icon: 'HiPencil',
    color: 'green',
  },
  {
    label: 'Delete',
    onButtonClick: () => {
      alert('Delete')
    },
    icon: 'HiTrash',
    color: 'red',
  },
]

export const extraMenuItems: MenuItemConfig[] = [
  {
    label: 'Go back',
    onButtonClick: () => {
      alert('Go back')
    },
  },
  {
    label: 'Create',
    onButtonClick: () => {
      alert('Create')
    },
  },
  {
    label: 'Move',
    onButtonClick: () => {
      alert('Move')
    },
  },
  {
    label: 'Close',
    onButtonClick: () => {
      alert('Close')
    },
  },
]
