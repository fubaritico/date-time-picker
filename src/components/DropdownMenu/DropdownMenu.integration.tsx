import { Dispatch, FC, ReactNode, SetStateAction, useState } from 'react'

import { DropdownMenuProps } from './DropdownMenu'

interface IntegrationProps extends DropdownMenuProps {
  children: (props: {
    props: DropdownMenuProps
    currentValue?: string
    setCurrentValue: Dispatch<SetStateAction<string | undefined>>
  }) => ReactNode
}

const Integration: FC<IntegrationProps> = (props) => {
  const [currentValue, setCurrentValue] = useState(props.value)

  return (
    <div style={{ width: 200 }}>
      {props.children({ props, currentValue, setCurrentValue })}
    </div>
  )
}

export default Integration
