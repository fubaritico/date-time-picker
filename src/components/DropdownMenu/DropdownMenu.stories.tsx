import { Meta, StoryObj } from '@storybook/react'
import { withCenteredLayout } from '@storybook-decorators/withLayout'

import { COLORS } from '@constants'

import DropdownMenu from './DropdownMenu'
import Integration from './DropdownMenu.integration'
import {
  coloredMenuItems,
  extraMenuItems,
  longMenuItems,
  menuItems,
} from './mock'
import HiEllipsisVertical from '../../assets/svg/HiEllipsisVertical.svg'

import type { DropdownMenuProps } from './DropdownMenu'
import type { FC } from 'react'

const DropdownWrapper: FC<DropdownMenuProps> = (args) => {
  return (
    <Integration {...args}>
      {({ props, currentValue, setCurrentValue }) => (
        <DropdownMenu
          {...props}
          value={currentValue}
          onValueChange={setCurrentValue}
        />
      )}
    </Integration>
  )
}

const meta: Meta<typeof DropdownWrapper> = {
  title: 'Components/Dropdown',
  parameters: {
    layout: 'fullscreen',
  },
  argTypes: {
    color: {
      options: COLORS,
      control: {
        type: 'select',
        labels: COLORS.reduce<Record<UIColor, string>>(
          (acc, color) => {
            acc[color] = color

            return acc
          },
          {} as Record<UIColor, string>
        ),
      },
    },
    variant: {
      control: 'select',
      options: ['primary', 'secondary'],
      description: 'The appearance of the dropdown',
    },
    size: {
      control: 'select',
      options: ['sm', 'md', 'lg'],
      description: 'The size of the dropdown',
    },
    placement: {
      control: 'select',
      options: ['bottom-start', 'bottom-end'],
      description: 'The placement of the dropdown menu',
    },
    defaultValue: {
      control: 'text',
      description: 'The default value of the Dropdown',
    },
  },
  tags: ['autodocs'],
  decorators: [withCenteredLayout],
  component: DropdownWrapper,
}
export default meta

type Story = StoryObj<typeof DropdownWrapper>

export const DropdownLeftAlignment: Story = {
  name: 'Blue with bottom-start placement',
  args: {
    color: 'blue',
    variant: 'primary',
    size: 'lg',
    placement: 'bottom-start',
    items: menuItems,
    menuWidth: 300,
    buttonLabel: 'Select an action',
  },
}

export const Primary: Story = {
  name: 'Primary with bottom-end placement',
  args: {
    variant: 'primary',
    size: 'lg',
    placement: 'bottom-end',
    items: menuItems,
    menuWidth: 280,
    buttonLabel: 'Select an action',
  },
}

export const WithLongItems: Story = {
  name: 'With long items',
  args: {
    variant: 'primary',
    size: 'lg',
    placement: 'bottom-end',
    items: longMenuItems,
    menuWidth: 280,
    buttonLabel: 'Select an action',
    showSelectedValue: true,
  },
}

export const WithManyItems: Story = {
  name: 'Menu with many items',
  args: {
    variant: 'primary',
    size: 'lg',
    placement: 'bottom-end',
    items: [...menuItems, ...extraMenuItems],
    menuWidth: 280,
    buttonLabel: 'Select an action',
  },
}

export const PrimaryWithColoredItemAndIcons: Story = {
  name: 'Colored menu items with icons',
  args: {
    variant: 'primary',
    size: 'lg',
    placement: 'bottom-start',
    items: coloredMenuItems,
    menuWidth: 280,
    buttonLabel: 'Select an action',
  },
}

export const Selected: Story = {
  name: 'Primary with selected value',
  args: {
    variant: 'primary',
    size: 'lg',
    defaultValue: 'Reassign',
    placement: 'bottom-end',
    items: menuItems,
    menuWidth: 280,
    buttonLabel: 'Select an action',
  },
}

export const SelectedShown: Story = {
  name: 'Primary with shown selected value',
  args: {
    variant: 'primary',
    size: 'lg',
    defaultValue: 'Reassign',
    placement: 'bottom-end',
    items: menuItems,
    menuWidth: 280,
    buttonLabel: 'Select an action',
    showSelectedValue: true,
  },
}

export const WithEmailAndName: Story = {
  name: 'Primary with name and email',
  args: {
    variant: 'primary',
    size: 'lg',
    placement: 'bottom-end',
    items: [
      {
        label: (
          <div className="react-node-example">
            <span>John Doe</span>
            <span>j.doe@odaseva.com</span>
          </div>
        ),
      },
      {
        separator: true,
      },
      ...menuItems,
    ],
    menuWidth: 280,
    buttonLabel: 'Select an action',
    showSelectedValue: true,
  },
}

export const IconButtonWithMenuToTheLeft: Story = {
  name: 'Icon button with menu to the left',
  args: {
    variant: 'secondary',
    placement: 'bottom-start',
    items: menuItems,
    menuWidth: 200,
    triggerComponent: (
      <div className="trigger-component-example">
        <HiEllipsisVertical />
      </div>
    ),
    showSelectedValue: true,
    buttonLabel: 'Select an action',
  },
}

export const IconButtonWithMenuToTheRight: Story = {
  name: 'Icon button with menu to the right',
  args: {
    ...IconButtonWithMenuToTheLeft.args,
    placement: 'bottom-end',
  },
}
