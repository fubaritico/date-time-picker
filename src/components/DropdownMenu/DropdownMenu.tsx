import clsx from 'clsx'
import { useCallback, useEffect, useRef, useState } from 'react'

import Button from '../Button'
import Label from '../Label'
import Menu from '../Menu'

import { isString } from './DropdownMenu.utils'
import { ReactComponent as HiChevronDown } from '../../assets/svg/HiChevronDown.svg'
import { ReactComponent as HiChevronUp } from '../../assets/svg/HiChevronUp.svg'

import type { ButtonProps, MenuItemConfig } from '@components'
import type { FC, ReactNode } from 'react'

export interface DropdownMenuProps {
  /* Extra CSS classes for the button (tailwind) */
  buttonClassName?: string
  /* The version of the button used in place of SimpleButton, with or without inactivity reset time (use for menu items) */
  buttonComponent?: typeof Button
  /* Default label displayed in the open button */
  buttonLabel?: ReactNode
  /* Button type */
  buttonType?: 'button' | 'submit' | 'reset'
  /* Button color */
  color?: UIColor
  /* Deprecated: Value on init (preselection in the case of a controlled component) */
  defaultValue?: string
  /* Dropdown is disabled */
  disabled?: boolean
  /* If true will render the menu in a portal */
  enablePortal?: boolean
  /* Array of menu items */
  items?: MenuItemConfig[]
  /* Text input label */
  label?: string
  /* Information displayed on hovering the exclamation mark in a tooltip */
  labelInfo?: string
  /* Loading state (in case values are fetched) */
  loading?: boolean
  /* Menu width in pixels */
  menuWidth?: number
  /* If true, the menu will have the same width as the trigger */
  menuFullWidth?: boolean
  /* If true will render the dropdown in full width */
  dropdownFullWidth?: boolean
  /* Callback called on any menu item click */
  onValueChange?: (value: string) => void
  /* The open state can be controlled */
  open?: boolean
  /* Menu placement: 'bottom-start' | 'bottom-end' */
  placement?: Extract<Placement, 'bottom-start' | 'bottom-end'>
  /* Dropdown is required */
  required?: boolean
  /* Extra CSS classes for the root element (tailwind) */
  rootClassName?: string
  /*
   * Will show the selected value instead of the button label, defaults to false.
   * Use this if you want a selectable menu item to be highlighted when selected.
   */
  showSelectedValue?: boolean
  /* Dropdown button size 'sm' | 'md' | 'lg' */
  size?: UISize
  /* Component used instead of the button, defaults to null */
  triggerComponent?: ReactNode
  /* Extra CSS classes for the trigger component element (tailwind) */
  triggerComponentClassName?: string
  /* Value passed to the component when controlled */
  value?: string
  /* Menu button style 'primary' | 'secondary' | 'ghost' */
  variant?: ButtonProps['variant']
  /* Width of the dropdown button in pixels */
  width?: number
}

const DropdownMenu: FC<DropdownMenuProps> = ({
  buttonComponent,
  buttonLabel,
  buttonType,
  color = 'stone',
  disabled,
  enablePortal = false,
  items,
  label,
  labelInfo,
  loading,
  menuWidth,
  menuFullWidth,
  placement = 'bottom-start',
  dropdownFullWidth = false,
  onValueChange,
  open,
  required,
  rootClassName,
  showSelectedValue,
  size = 'lg',
  triggerComponent = null,
  triggerComponentClassName,
  value,
  variant = 'primary',
  width,
}) => {
  const [isOpen, setIsOpen] = useState(open ?? false)
  const buttonContainerRef = useRef<HTMLDivElement>(null)
  const currentLabel =
    showSelectedValue && value
      ? items?.find((i) => i.value === value)?.label
      : (buttonLabel ?? '')

  /**
   *
   */
  const closeMenu = useCallback(() => {
    setIsOpen(false)
  }, [])

  /**
   * Will emit the value of a selectable menu item
   *
   * @param item
   */
  const handleMenuItemClick = (item: MenuItemConfig) => {
    if (item.type === 'selectable' && !item.value) {
      throw Error('Value is required for selectable menu item')
    }

    if (isString(item.label) && item.type === 'selectable' && !!item.value) {
      onValueChange?.(item.value)
    }

    if (item.type !== 'selectable') {
      // Will unselect any selectable menu item already selected
      onValueChange?.('')
    }

    closeMenu()
  }

  useEffect(() => {
    document
      .getElementById('page-content')
      ?.addEventListener('scroll', closeMenu)

    return () => {
      document
        .getElementById('page-content')
        ?.removeEventListener('scroll', closeMenu)
    }
  }, [closeMenu])

  return (
    <div
      className={clsx(
        'Dropdown',
        {
          'full-width': !dropdownFullWidth,
          'menu-full-width': menuFullWidth,
          disabled: disabled ?? loading,
        },
        rootClassName
      )}
      style={{ width }}
    >
      {label && (
        <Label
          label={label}
          labelInfo={labelInfo}
          required={required}
          size={size}
          disabled={disabled}
        />
      )}
      <div className="container">
        <div
          className={clsx(
            'button-container',
            {
              'has-trigger-component': triggerComponent,
            },
            triggerComponentClassName
          )}
          ref={buttonContainerRef}
          style={{ width }}
        >
          {triggerComponent ? (
            <button
              onClick={() => {
                setIsOpen(!isOpen)
              }}
              aria-label={isString(buttonLabel) ? buttonLabel : undefined}
            >
              {triggerComponent}
            </button>
          ) : (
            <Button
              aria-label={
                isString(buttonLabel) ? buttonLabel : 'Select an option'
              }
              icon={isOpen ? HiChevronDown : HiChevronUp}
              iconPosition="right"
              data-test="dropdown-value"
              label={currentLabel}
              loading={loading}
              onClick={() => {
                setIsOpen(!isOpen)
              }}
              size={size}
              color={color}
              style={{ width }}
              type={buttonType}
              variant={variant}
            />
          )}
        </div>
        <Menu
          buttonComponent={buttonComponent}
          color={color}
          enablePortal={enablePortal}
          items={items ?? []}
          menuWidth={menuWidth}
          onClose={closeMenu}
          onMenuItemClick={handleMenuItemClick}
          open={isOpen}
          placement={placement}
          selectedValue={value}
          showSelectedValue={showSelectedValue}
          size={size}
          triggerRef={buttonContainerRef}
        />
      </div>
    </div>
  )
}

export default DropdownMenu
