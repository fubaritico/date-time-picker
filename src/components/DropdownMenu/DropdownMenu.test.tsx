import { act, render, screen, waitFor, within } from '@testing-library/react'
import user from '@testing-library/user-event'

import DropdownMenu, { DropdownMenuProps } from './DropdownMenu'
import Integration from './DropdownMenu.integration'
import { selectableMenuItems } from './mock'

import type { MenuItemConfig } from '@components'

export const actionsMenuItems: MenuItemConfig[] = [
  {
    label: 'Quick save ',
    onButtonClick: jest.fn(),
    type: 'action',
  },
  {
    label: 'Save',
    onButtonClick: jest.fn(),
    type: 'action',
  },
  {
    label: 'Reassign',
    onButtonClick: jest.fn(),
    type: 'action',
  },
  {
    label: 'Edit',
    onButtonClick: jest.fn(),
    type: 'action',
  },
  {
    label: 'Delete',
    onButtonClick: jest.fn(),
    type: 'action',
  },
]

const setupControlledDropdown = (args: DropdownMenuProps) => {
  const { container } = render(
    <Integration {...args}>
      {({ props, currentValue, setCurrentValue }) => (
        <DropdownMenu
          {...props}
          value={currentValue}
          onValueChange={setCurrentValue}
        />
      )}
    </Integration>
  )

  return { container }
}

const setupDropdown = (props: DropdownMenuProps) => {
  const { container } = render(<DropdownMenu {...props} />)

  return { container }
}

describe('DropdownMenu', () => {
  describe('Selectable', () => {
    const initialProps: DropdownMenuProps = {
      items: selectableMenuItems,
      buttonLabel: 'Select a value',
    }

    it('should render', () => {
      const { container } = setupControlledDropdown(initialProps)

      expect(container).toBeInTheDocument()
      expect(
        screen.getByRole('button', { name: initialProps.buttonLabel as string })
      ).toBeInTheDocument()

      expect(container).toMatchSnapshot()
    })

    it('should render menu on label/button click', async () => {
      const { container } = setupControlledDropdown(initialProps)

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      expect(await screen.findByTestId('dropdown-menu')).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    // triggers warning
    it('should not render menu on document/body click', async () => {
      setupControlledDropdown(initialProps)

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      expect(await screen.findByTestId('dropdown-menu')).toBeInTheDocument()

      await act(async () => {
        await user.click(document.body)
      })

      await waitFor(() => {
        expect(screen.queryByTestId('dropdown-menu')).not.toBeInTheDocument()
      })
    })

    // triggers warning
    it('should hide the dropdown list on item selection/click', async () => {
      setupControlledDropdown(initialProps)

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      expect(await screen.findByTestId('dropdown-menu')).toBeInTheDocument()

      await act(async () => {
        await user.click(
          within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
            hidden: true,
          })[0]
        )
      })

      // triggers warning if a disappearance is asserted (transition seems tobe involved)
      // await waitForElementToBeRemoved(() =>
      //   screen.queryByTestId('dropdown-menu')
      // )
      await waitFor(() => {
        expect(screen.queryByTestId('dropdown-menu')).not.toBeInTheDocument()
      })
    })

    // triggers warning
    it('should display selected value when "showSelectedValue" is true on item selection/click', async () => {
      const showSelectedValueProps: DropdownMenuProps = {
        ...initialProps,
        showSelectedValue: true,
      }

      setupControlledDropdown(showSelectedValueProps)

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      expect(await screen.findByTestId('dropdown-menu')).toBeInTheDocument()

      await act(async () => {
        await user.click(
          within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
            hidden: true,
          })[0]
        )
      })

      await waitFor(() => {
        expect(screen.queryByTestId('dropdown-menu')).not.toBeInTheDocument()
      })

      expect(
        screen.getByRole('button', {
          name: initialProps.buttonLabel as string,
        })
      ).toHaveTextContent((selectableMenuItems[0].label as string).trim())
    })

    it('should display menu item as selected after selection when "showSelectedValue" is true', async () => {
      const showSelectedValueProps: DropdownMenuProps = {
        ...initialProps,
        showSelectedValue: true,
      }

      setupControlledDropdown(showSelectedValueProps)

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      await act(async () => {
        await user.click(
          within(await screen.findByTestId('dropdown-menu')).getAllByRole(
            'button',
            {
              hidden: true,
            }
          )[0]
        )
      })

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      expect(
        within(screen.getByTestId('dropdown-menu')).getAllByRole('listitem', {
          hidden: true,
        })[0]
      ).toHaveClass('selected')

      expect(
        within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
          hidden: true,
        })[0]
      ).toHaveAttribute('aria-current', 'true')
    })

    it('should display menu item as selected and selected value in label placeholder after if value is passed when "showSelectedValue" is true', async () => {
      const showSelectedValueProps: DropdownMenuProps = {
        ...initialProps,
        showSelectedValue: true,
        value: selectableMenuItems[1].value,
      }

      setupControlledDropdown(showSelectedValueProps)

      expect(
        screen.getByText(selectableMenuItems[1].label as string)
      ).toBeInTheDocument()

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: 'Select a value',
          })
        )
      })

      expect(
        within(screen.getByTestId('dropdown-menu')).getAllByRole('listitem', {
          hidden: true,
        })[1]
      ).toHaveClass('selected')

      expect(
        within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
          hidden: true,
        })[1]
      ).toHaveAttribute('aria-current', 'true')
    })
  })

  describe('Actions', () => {
    const initialProps: DropdownMenuProps = {
      items: actionsMenuItems,
      buttonLabel: 'Select an action',
    }

    it('should call action function on menu item click', async () => {
      setupDropdown(initialProps)

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      expect(await screen.findByTestId('dropdown-menu')).toBeInTheDocument()

      await act(async () => {
        await user.click(
          within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
            hidden: true,
          })[0]
        )
      })

      await waitFor(() => {
        expect(screen.queryByTestId('dropdown-menu')).not.toBeInTheDocument()
      })

      expect(actionsMenuItems[0].onButtonClick).toHaveBeenCalledTimes(1)
    })

    it('should not change display state of the dropdown after action click (label, selection)', async () => {
      setupDropdown(initialProps)

      await act(async () => {
        await user.click(
          screen.getByRole('button', {
            name: initialProps.buttonLabel as string,
          })
        )
      })

      await act(async () => {
        await user.click(
          within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
            hidden: true,
          })[0]
        )
      })

      await waitFor(() => {
        expect(screen.queryByTestId('dropdown-menu')).not.toBeInTheDocument()
      })

      expect(
        screen.getByText(initialProps.buttonLabel as string)
      ).toBeInTheDocument()

      await act(async () => {
        await user.click(screen.getByText(initialProps.buttonLabel as string))
      })

      expect(
        within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
          hidden: true,
        })[0]
      ).not.toHaveClass('!bg-blue-100 hover:!bg-blue-200 !text-blue-500')

      expect(
        within(screen.getByTestId('dropdown-menu')).getAllByRole('button', {
          hidden: true,
        })[0]
      ).toHaveAttribute('aria-current', 'false')
    })
  })
})
