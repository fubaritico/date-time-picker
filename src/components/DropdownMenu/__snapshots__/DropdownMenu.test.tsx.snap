// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu Selectable should render 1`] = `
<div>
  <div
    style="width: 200px;"
  >
    <div
      class="Dropdown full-width"
    >
      <div
        class="container"
      >
        <div
          class="button-container"
        >
          <button
            aria-label="Select a value"
            class="Button lg primary stone right has-icon"
            data-test="dropdown-value"
          >
            <svg
              data-jest-file-name="HiChevronUp.svg"
              data-jest-svg-name="HiChevronUp"
              data-testid="HiChevronUp"
            />
            <span>
              Select a value
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownMenu Selectable should render menu on label/button click 1`] = `
<div>
  <div
    style="width: 200px;"
  >
    <div
      class="Dropdown full-width"
    >
      <div
        class="container"
      >
        <div
          class="button-container"
        >
          <button
            aria-label="Select a value"
            class="Button lg primary stone right has-icon"
            data-test="dropdown-value"
          >
            <svg
              data-jest-file-name="HiChevronDown.svg"
              data-jest-svg-name="HiChevronDown"
              data-testid="HiChevronDown"
            />
            <span>
              Select a value
            </span>
          </button>
        </div>
        <div>
          <div
            class="Menu lg menu-enter menu-enter-active"
            data-test="dropdown-menu"
            style="position: absolute; top: 0px; left: 0px; visibility: hidden;"
          >
            <ul>
              <li
                class="MenuItem"
              >
                <button
                  aria-disabled="false"
                  aria-label="Quick save "
                  aria-selected="false"
                  class="Button lg ghost stone left not-rounded"
                >
                  <span>
                    <span
                      id="item-Quick save"
                    >
                      Quick save 
                    </span>
                  </span>
                </button>
              </li>
              <li
                class="MenuItem"
              >
                <button
                  aria-disabled="false"
                  aria-label="Save"
                  aria-selected="false"
                  class="Button lg ghost stone left not-rounded"
                >
                  <span>
                    <span
                      id="item-Save"
                    >
                      Save
                    </span>
                  </span>
                </button>
              </li>
              <li
                class="MenuItem"
              >
                <button
                  aria-disabled="false"
                  aria-label="Reassign"
                  aria-selected="false"
                  class="Button lg ghost stone left not-rounded"
                >
                  <span>
                    <span
                      id="item-Reassign"
                    >
                      Reassign
                    </span>
                  </span>
                </button>
              </li>
              <li
                class="MenuItem"
              >
                <button
                  aria-disabled="false"
                  aria-label="Edit"
                  aria-selected="false"
                  class="Button lg ghost stone left not-rounded"
                >
                  <span>
                    <span
                      id="item-Edit"
                    >
                      Edit
                    </span>
                  </span>
                </button>
              </li>
              <li
                class="MenuItem"
              >
                <button
                  aria-disabled="false"
                  aria-label="Delete"
                  aria-selected="false"
                  class="Button lg ghost stone left not-rounded"
                >
                  <span>
                    <span
                      id="item-Delete"
                    >
                      Delete
                    </span>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
